rllib/algorithms/appo/tf/appo_tf_learner.py

```diff
@@ -215,38 +217,38 @@ def remove_module(self, module_id: str):
         super().remove_module(module_id)
         self.kl_coeffs.pop(module_id)
 
-    def _update_module_target_pi(self):
+    def _update_module_target_networks(self, module_id: ModuleID):
         """Update the target policy of each module with the current policy.
 
         Do that update via polyak averaging.
 
+        Args:
+            module_id: The module whose target networks need to be updated.
+
         """
-        module_ids = self.module.keys()
-        for module_id in module_ids:
-            module = self.module[module_id]
-            old_encoder = module.old_encoder
-            old_pi = module.old_pi
-            current_encoder = module.encoder
-            current_pi = module.pi
-
-            encoder_old_current_pair = {"old": old_encoder, "current": current_encoder}
-            pi_old_current_pair = {"old": old_pi, "current": current_pi}
-            for network_pair in [encoder_old_current_pair, pi_old_current_pair]:
-                for old_var, current_var in zip(
-                    network_pair["old"].variables, network_pair["current"].variables
-                ):
-                    updated_var = self.tau * current_var + (1.0 - self.tau) * old_var
-                    old_var.assign(updated_var)
-
-    def _update_module_kl_coeff(self):
+        module = self.module[module_id]
+
+        target_current_network_pairs = module.target_networks()
+        for target_network, current_network in target_current_network_pairs:
+            for old_var, current_var in zip(
+                target_network.variables, current_network.variables
+            ):
+                updated_var = self.tau * current_var + (1.0 - self.tau) * old_var
+                old_var.assign(updated_var)
+
+    def _update_module_kl_coeff(self, module_id: ModuleID):
         """Dynamically update the KL loss coefficients of each module with.
 
         The update is completed using the mean KL divergence between the action
         distributions current policy and old policy of each module. That action
         distribution is computed during the most recent update/call to `compute_loss`.
 
+        Args:
+            module_id: The module whose KL loss coefficient to update.
+
         """
-        for module_id, sampled_kl in self.sampled_kls.items():
+        if module_id in self.sampled_kls:
```

Comment:
I just noticed that sampled_kls is a state variable for the learner, but this will lead to information leakage. Where is the information leakage? 
Well you should implicitly know that compile_results() will fill in this state variable so that you can run the additional_update(). 

The correct way to do it is to retrieve sampled kl values from the results and then pass them as kwargs to the additional updates method. Look at PPOs implementation for this detail. We have the same concept there, but inconsistent implementation between PPO and APPO.

Reply:
fixed